File: past_archive/boj/14503(로봇청소기).py
from collections import deque
ans=0
N,M=map(int,input().split())
r,c,d=map(int,input().split())
L=[]
for i in range(N):
    L.append(list(map(int,input().split())))

dr=[-1,0,1,0]
dc=[0,1,0,-1]

backr=[1,0,-1,0]
backc=[0,-1,0,1]

if L[r][c]==0:
    L[r][c]=1
    ans+=1

while(1):
    tempD=d
    isClean=False
    for i in range(4):
        tempD-=1
        if tempD<0:
            tempD+=4
        tempR=r+dr[tempD]
        tempC=c+dc[tempD]
        if L[tempR][tempC]==0 :
            ans+=1
            isClean=True
            L[tempR][tempC]=3
            r=tempR
            c=tempC
            d=tempD
            break

    #청소실패
    if not isClean:
        r=r+backr[d]
        c=c+backc[d]
        if L[r][c]==1 :
            break



print(ans)

File: past_archive/boj/14503(로봇청소기).cpp
#include <iostream>
using namespace std;

#define maxSize 51

int main() {
	int ans = 1;
	int N, M, r, c, d;
	cin >> N >> M;
	cin >> r >> c >> d;
	int L[maxSize][maxSize];
	int dr[4][4] = { {0,1,0,-1},{-1,0,1,0},{0,-1,0,1},{1,0,-1,0} };
	int dc[4][4] = { {-1,0,1,0},{0,-1,0,1},{1,0,-1,0},{0,1,0,-1} };

	int backR[4] = { 1,0,-1,0 };
	int backC[4] = { 0,-1,0,1 };

	for (int i = 0; i < N; i++) {
		for (int j = 0; j < M; j++) {
			cin >> L[i][j];
		}
	}
	L[r][c] = 3;

	while (1) {
		/*
		cout << r << c << d << endl;
		for (int i = 0; i < N; i++) {
			for (int j = 0; j < M; j++) {
				cout << L[i][j] << " ";
			}
			cout << endl;
		}
		*/

		bool isClean = false;
		for (int i = 0; i < 4; i++) {
			int tempR = r + dr[d][i];
			int tempC = c + dc[d][i];
			int tempD = d - (i + 1);

			if (tempD < 0) {
				tempD += 4;
			}

			if (L[tempR][tempC] == 0) {
				isClean = true;
				ans += 1;
				L[tempR][tempC] = 3;
				r = tempR;
				c = tempC;
				d = tempD;
				break;
			}
		}
		if (!isClean) {
			r = r + backR[d];
			c = c + backC[d];
			if (L[r][c] == 1) {
				break;
			}
		}
		

	}
	cout << ans;
	return 0;
}